astropy/units/decorators.py

```diff
@@ -127,7 +139,11 @@ def wrapper(*func_args, **func_kwargs):
 
             # Call the original function with any equivalencies in force.
             with add_enabled_equivalencies(self.equivalencies):
-                return wrapped_function(*func_args, **func_kwargs)
+                return_ = wrapped_function(*func_args, **func_kwargs)
+            if isinstance(wrapped_signature.return_annotation, (Unit, Quantity)):
```

Comment:
I have a slight worry here, because function units such as `u.mag(u.Jy)` or `u.STmag` are *not* `Unit` instances (this does need to be fixed eventually; #3650). My sense is that if one uses `@quantity_input`, then one should not really expect to be able to use the output annotation for anything else, so I think I would just stick with what you had. But if you do want to be more general, I would suggest:
```
try:
    return_unit = Unit(wrapped_signature.return_annotation)
except Exception:  # not a unit requested, so ignore
    pass
else:
    return return_.to(return_unit)
```
Obviously, the above, like your original implementation, allows one to provide units as strings, which I think is OK.